build_tools/memory_monitor.py

```diff
@@ -0,0 +1,548 @@
+from pathlib import Path
+from typing import Optional, Dict, Any
+import signal
+import threading
+import psutil
+
+sys.path.insert(0, str(Path(__file__).parent))
+from github_actions.github_actions_utils import gha_append_step_summary
+
+
+class MemoryMonitor:
+    """Monitors system and process memory usage."""
+
+    def __init__(
+        self,
+        interval_seconds: float = 5.0,
+        phase_name: str = "Unknown",
+        log_file: Optional[Path] = None,
+        stop_signal_file: Optional[Path] = None,
+    ):
+        self.interval_seconds = interval_seconds
+        self.phase_name = phase_name
+        self.log_file = log_file
+        self.stop_signal_file = stop_signal_file
+        self.stop_event = threading.Event()
+        self.peak_memory = 0
+        self.peak_swap = 0
+        self.samples = []
+        self.start_time = None
+        self.end_time = None
+
+    def get_memory_stats(self) -> Dict[str, Any]:
+        """Collect current memory statistics."""
+        vm = psutil.virtual_memory()
+        swap = psutil.swap_memory()
+
+        # Get current process and its children
+        current_process = psutil.Process()
+        process_memory = current_process.memory_info().rss
+
+        # Try to get memory of all child processes
+        children_memory = 0
+        try:
+            children = current_process.children(recursive=True)
+            for child in children:
+                try:
+                    children_memory += child.memory_info().rss
+                except (
+                    psutil.NoSuchProcess,
+                    psutil.AccessDenied,
+                    psutil.ZombieProcess,
+                ):
+                    pass
+        except (psutil.NoSuchProcess, psutil.AccessDenied, psutil.ZombieProcess):
+            pass
+
+        total_process_memory = process_memory + children_memory
+
+        # Track peak usage
+        self.peak_memory = max(self.peak_memory, vm.used)
+        self.peak_swap = max(self.peak_swap, swap.used)
+
+        stats = {
+            "timestamp": datetime.now().isoformat(),
+            "phase": self.phase_name,
+            # System memory
+            "total_memory_gb": vm.total / (1024**3),
+            "available_memory_gb": vm.available / (1024**3),
+            "used_memory_gb": vm.used / (1024**3),
+            "memory_percent": vm.percent,
+            "free_memory_gb": vm.free / (1024**3),
+            # Peak memory
+            "peak_memory_gb": self.peak_memory / (1024**3),
+            "peak_swap_gb": self.peak_swap / (1024**3),
+            # Swap
+            "total_swap_gb": swap.total / (1024**3),
+            "used_swap_gb": swap.used / (1024**3),
+            "swap_percent": swap.percent,
+            # Process memory
+            "process_memory_gb": process_memory / (1024**3),
+            "children_memory_gb": children_memory / (1024**3),
+            "total_process_memory_gb": total_process_memory / (1024**3),
+        }
```

Comment:
This is collecting similar statistics to https://github.com/ROCm/TheRock/pull/2475. Do we have a design document, tracking issue, RFC, or some other area where we've written down what metrics we want to collect and have considered how to collect those metrics across relevant jobs?